### PR TITLE
Photo Sharing Sample App with UI tests. Python script to build and run tests.

### PR DESCRIPTION
**Pull request by changxu0306**
Description of this PR:
The whole sample app code is committed to this PR, with Auth, AppSync and Storage implemented.
- Define Album class and Photo class for album and photo objects. Make it clear for Object-Oriented Programming.
- AppSyncHelper class provides AppSync utilities for the sample app. When an AppSyncHelper object is created, it will initialize AWS AppSyncClient with Cognito User Pools by its constructor. Within the class, we provide multiple queries and mutations (create, update, delete) for albums and photos.
- StorageHelper class provides Storage utilities (TransferUtility) for the sample app. It also provides useful type convertion APIs between bitmaps, files, drawables and URIs.
- AlbumActivity view displays all the albums belong to the user who signed in. User can create an album, delete an album, update an album's name or sign out in this activity.
- PhotoActivity view allows user upload an image file and download image files from Amazon S3 bucket. After download each image file, it will show up on the screen in a grid view.
- DownloadSelectionActivity view shows a list of items which are stored in the S3 bucket. User can choose one file at a time to download by clicking on an item of the list. After clicking, it will go back to PhotoActivity.
- AlbumAdapter and PhotoAdapter are customized adapters which bind attributes of an Album or Photo object with a UI item. Listeners are defined when clicking add button and delete button. It allows UI changes dynamically with the latest view.

### CODE REVIEW

**Comment by changxu0306** [comment]
Add python scripts to build and run UI tests automatically.
test_names.json stores the class names of UI tests.

File: PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/AppSyncHelper.java [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    private GraphQLCall.Callback<CreateAlbumMutation.Data> createAlbumCallback = new GraphQLCall.Callback<CreateAlbumMutation.Data>() {
+        @Override
+        public void onResponse(@Nonnull Response<CreateAlbumMutation.Data> response) {
+            String albumName = response.data().createAlbum().name();
+            Log.i(TAG, "An album with name" + albumName + "is created successfully.");
+        }
+
+        @Override
+        public void onFailure(@Nonnull ApolloException e) {
+            Log.i(TAG, "Failed to create an album.");
```
Comment (kvasukib) [inline PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/AppSyncHelper.java]: Add an AlertDialog when there is a failure to report to the user of the sample app. Repeat this for all success and failure scenarios.

File: PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/AppSyncHelper.java [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+     */
+    public ArrayList<Album> listAlbums() {
+        final ArrayList<Album> albumList = new ArrayList<Album>();
+
+        final CountDownLatch listAlbumsLatch = new CountDownLatch(1);
+        // limit is set to ensure the all the albums can be queried successfully if the number of albums is less than or equal to 30.
+        // The number of limit can be adjusted by customers.
+
+        // Using NETWORK_ONLY state is set to AppSyncResponseFetcher in case deleted albums still stay in cache.
+        mAWSAppSyncClient.query(ListAlbumsQuery.builder().limit(30).build()).responseFetcher(AppSyncResponseFetchers.NETWORK_ONLY)
```
Comment (kvasukib) [inline PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/AppSyncHelper.java]: Add comment on why 30.

File: PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/Photo.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    // photo file name
+    private String name;
+    // the bucket which the photo stores in
+    private String bucket;
+    // key in the bucket
+    private String key;
+    // photo file in Bitmap format
+    private Bitmap file;
+
+    // Constructor
```
Comment (kvasukib) [inline PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/Photo.java]: Add Javadoc

File: PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/PhotoActivity.java [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+     * Define actions on the menu.
+     * @param item
+     * @return
+     */
+    @Override
+    public boolean onOptionsItemSelected(MenuItem item) {
+        switch (item.getItemId()) {
+            case R.id.photo_add:
+                // this is for user to choose an image file from device.
+//                Intent intent = new Intent();
```
Comment (kvasukib) [inline PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/PhotoActivity.java]: Remove the commented code if not necessary.

File: PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/PhotoActivity.java [hunk 5]
```diff
@@ -0,0 +1,523 @@
...
+        final String filename = file.getName();
+        final String key = getKeyInBucket(filename, accessType);
+        TransferObserver observer = transferUtility.upload(key, file);
+
+        // Attach a listener to the observer to get state update and progress notifications
+        observer.setTransferListener(new TransferListener() {
+            @Override
+            public void onStateChanged(int id, TransferState state) {
+                if (state.equals(TransferState.COMPLETED)) {
+                    Toast.makeText(PhotoActivity.this, "Upload succeed!",
```
Comment (kvasukib) [inline PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/PhotoActivity.java]: Remove the toast as alert dialog serves that purpose.

File: PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/PhotoActivity.java [hunk 6]
```diff
@@ -0,0 +1,523 @@
...
+            }
+
+            @Override
+            public void onProgressChanged(int id, long bytesCurrent, long bytesTotal) {
+                Log.d(TAG, "Progress = " + bytesCurrent + " / " + bytesTotal);
+            }
+
+            @Override
+            public void onError(int id, Exception ex) {
+
```
Comment (kvasukib) [inline PhotoAlbumSample/app/src/main/java/com/amazonaws/android/samples/photosharing/PhotoActivity.java]: log the exception.

File: PhotoAlbumSample/app/src/main/res/raw/awsconfiguration.json [hunk 7]
```diff
@@ -0,0 +1,37 @@
+{
```
Comment (kvasukib) [inline PhotoAlbumSample/app/src/main/res/raw/awsconfiguration.json]: You should not be committing this file. Please remove it immediately as it contains sensitive information.

**Pull request closed without merging**